rust/src/bs_safe.rs

```diff
@@ -0,0 +1,535 @@
+            n,
+            "Argument 'theta_unc' must be the same size as the number of parameters!"
+        );
+        assert_eq!(
+            grad.len(),
+            n,
+            "Argument 'grad' must be the same size as the number of parameters!"
+        );
+        assert_eq!(
+            hessian.len(),
+            n.checked_mul(n).expect("Overflow for size of hessian"),
+            "Argument 'hessian' must be the same size as the number of parameters squared!"
+        );
+
+        let mut val = 0.0;
+
+        let mut err = ErrorMsg::new(self.lib.borrow());
+        let rc = unsafe {
+            self.lib.borrow().0.bs_log_density_hessian(
+                self.model.as_ptr(),
+                propto as c_int,
+                jacobian as c_int,
+                theta_unc.as_ptr(),
+                &mut val,
+                grad.as_mut_ptr(),
+                hessian.as_mut_ptr(),
+                err.as_ptr(),
+            )
+        };
+
+        if rc == 0 {
+            Ok(val)
+        } else {
+            Err(BridgeStanError::EvaluationFailed(err.message()))
+        }
+    }
+
+    /// Map a point in unconstrained parameter space to the constrained space
+    pub fn param_constrain<R: Borrow<StanLibrary>>(
+        &self,
+        theta_unc: &[f64],
+        include_tp: bool,
+        include_gq: bool,
+        out: &mut [f64],
+        rng: &mut Rng<R>,
+    ) -> Result<()> {
+        let n = self.param_unc_num();
+        assert_eq!(
+            theta_unc.len(),
+            n,
+            "Argument 'theta_unc' must be the same size as the number of parameters!"
+        );
+        let out_n = self.param_num(include_tp, include_gq);
+        assert_eq!(
+            out.len(),
+            out_n,
+            "Argument 'out' must be the same size as the number of parameters!"
+        );
+
+        let mut err = ErrorMsg::new(self.lib.borrow());
+        let rc = unsafe {
+            self.lib.borrow().0.bs_param_constrain(
+                self.model.as_ptr(),
+                include_tp as c_int,
+                include_gq as c_int,
+                theta_unc.as_ptr(),
+                out.as_mut_ptr(),
+                rng.rng.as_ptr(),
+                err.as_ptr(),
+            )
+        };
+
+        if rc == 0 {
+            Ok(())
+        } else {
+            Err(BridgeStanError::EvaluationFailed(err.message()))
+        }
+    }
+
+    /// Set the sequence of unconstrained parameters based on the
+    /// specified constrained parameters, and return a return code of 0
+    /// for success and -1 for failure.  Parameter order is as declared
+    /// in the Stan program, with multivariate parameters given in
+    /// last-index-major order.
+    ///
+    /// # Arguments
+    /// `theta`: The vector of constrained parameters
+    ///
+    /// `theta_unc` Vector of unconstrained parameters
+    pub fn param_unconstrain(&self, theta: &mut [f64], theta_unc: &mut [f64]) -> Result<()> {
```

Comment:
Missing the `param_unconstrain_json` function following this one

Reply:
done